File: Polars.FSharp/Enums.fs
namespace Polars.FSharp

open Polars.NET.Core

type TimeUnit = 
    | Nanoseconds
    | Microseconds
    | Milliseconds

type Field = { Name: string; DataType: DataType }

/// <summary>
/// Polars data types for casting and schema definitions.
/// </summary>
and DataType =
    | Boolean
    | Int8 | Int16 | Int32 | Int64
    | UInt8 | UInt16 | UInt32 | UInt64
    | Float32 | Float64
    | String
    | Date | Datetime of TimeUnit * string option | Time
    | Duration of TimeUnit
    | Binary
    | Categorical
    | Decimal of precision: int option * scale: int option
    | Unknown | SameAsInput | Null | List of DataType | Array of DataType * width: uint64
    | Struct of Field list 
    | Int128 | UInt128
    member this.Code : int =
        match this with
        | Unknown | SameAsInput -> 0
        | Boolean -> 1
        | Int8 -> 2
        | Int16 -> 3
        | Int32 -> 4
        | Int64 -> 5
        | UInt8 -> 6
        | UInt16 -> 7
        | UInt32 -> 8
        | UInt64 -> 9
        | Float32 -> 10
        | Float64 -> 11
        | String -> 12
        | Date -> 13
        | Datetime _ -> 14 
        | Time -> 15
        | Duration _ -> 16
        | Binary -> 17
        | Null -> 18
        | Struct _ -> 19
        | List _ -> 20
        | Categorical -> 21
        | Decimal _ -> 22
        | Array _ -> 23
        | Int128 -> 24
        | UInt128 -> 25
    static member FromHandle (handle: DataTypeHandle) : DataType =
        let kind = PolarsWrapper.GetDataTypeKind handle

        match kind with
        | 1 -> Boolean
        | 2 -> Int8
        | 3 -> Int16
        | 4 -> Int32
        | 5 -> Int64
        | 6 -> UInt8
        | 7 -> UInt16
        | 8 -> UInt32
        | 9 -> UInt64
        | 10 -> Float32
        | 11 -> Float64
        | 12 -> String 
        | 13 -> Date
        | 24 -> Int128
        | 25 -> UInt128
        
        // --- Complex Type ---
        
        // Datetime
        | 14 -> 
            let unitCode = PolarsWrapper.GetTimeUnit handle
            let unit = 
                match unitCode with 
                | 0 -> Nanoseconds 
                | 1 -> Microseconds 
                | 2 -> Milliseconds 
                | _ -> Microseconds
            
            let tz = Option.ofObj (PolarsWrapper.GetTimeZone handle)
            Datetime(unit, tz)

        | 15 -> Time
        
        // Duration
        | 16 -> 
            let unitCode = PolarsWrapper.GetTimeUnit handle
            let unit = 
                match unitCode with 
                | 0 -> Nanoseconds 
                | 1 -> Microseconds 
                | 2 -> Milliseconds 
                | _ -> Microseconds
            Duration unit

        | 17 -> Binary
        | 18 -> Null
        
        // Struct
        | 19 -> 
            let len = PolarsWrapper.GetStructLen handle
            let fields = 
                [ for i in 0UL .. len - 1UL do
                    let mutable name = Unchecked.defaultof<string>
                    let mutable fieldHandle = Unchecked.defaultof<DataTypeHandle>
                    
                    PolarsWrapper.GetStructField(handle, i, &name, &fieldHandle)

                    use h = fieldHandle 
                    yield { Name = name; DataType = DataType.FromHandle h }
                ]
            Struct fields

        // List
        | 20 -> 
            use innerHandle = PolarsWrapper.GetInnerType handle
            let innerType = DataType.FromHandle innerHandle
            List innerType

        | 21 -> Categorical

        // Decimal
        | 22 -> 
            let mutable prec = 0
            let mutable scale = 0
            PolarsWrapper.GetDecimalInfo(handle, &prec, &scale)
            Decimal(Some prec,Some scale)

        | 23 -> 
            use innerHandle = PolarsWrapper.GetInnerType handle
            let width = PolarsWrapper.DataTypeGetArrayWidth handle;
            let innerType = DataType.FromHandle innerHandle
            Array(innerType,width)

        | _ -> Unknown

    member this.IsNumeric =
        match this with
        | UInt8 | UInt16 | UInt32 | UInt64
        | Int8 | Int16 | Int32 | Int64
        | Float32 | Float64 | Int128
        | Decimal _ -> true
        | _ -> false

    /// <summary>
    /// Creates a native Polars DataTypeHandle from this F# DataType.
    /// Recursive structures (List, Struct) are handled automatically.
    /// </summary>
    member internal this.CreateHandle() : DataTypeHandle =
        
        let toUnitCode tu = 
            match tu with 
            | Nanoseconds -> 0 
            | Microseconds -> 1 
            | Milliseconds -> 2

        match this with
        | SameAsInput -> PolarsWrapper.NewPrimitiveType 0
        | Null -> PolarsWrapper.NewPrimitiveType 18
        | Boolean -> PolarsWrapper.NewPrimitiveType 1
        | Int8 -> PolarsWrapper.NewPrimitiveType 2
        | Int16 -> PolarsWrapper.NewPrimitiveType 3
        | Int32 -> PolarsWrapper.NewPrimitiveType 4
        | Int64 -> PolarsWrapper.NewPrimitiveType 5
        | UInt8 -> PolarsWrapper.NewPrimitiveType 6
        | UInt16 -> PolarsWrapper.NewPrimitiveType 7
        | UInt32 -> PolarsWrapper.NewPrimitiveType 8
        | UInt64 -> PolarsWrapper.NewPrimitiveType 9
        | Float32 -> PolarsWrapper.NewPrimitiveType 10
        | Float64 -> PolarsWrapper.NewPrimitiveType 11
        | String -> PolarsWrapper.NewPrimitiveType 12
        | Binary -> PolarsWrapper.NewPrimitiveType 17
        | Date -> PolarsWrapper.NewPrimitiveType 13
        | Time -> PolarsWrapper.NewPrimitiveType 15
        | Int128 -> PolarsWrapper.NewPrimitiveType 24
        | UInt128 -> PolarsWrapper.NewPrimitiveType 25
        
        // --- Complex Type ---

        // Datetime: Unit and Timezone
        | Datetime(unit, tz) ->
            let code = toUnitCode unit
            let tzStr = Option.toObj tz // None -> null
            PolarsWrapper.NewDateTimeType(byte code, tzStr)

        // Duration
        | Duration unit ->
            let code = toUnitCode unit
            PolarsWrapper.NewDurationType (byte code)

        // Categorical
        | Categorical -> 
            PolarsWrapper.NewCategoricalType()

        // Decimal: precision (p, s)
        | Decimal(p, s) ->
            let prec = defaultArg p 0
            let scale = defaultArg s 0 
            PolarsWrapper.NewDecimalType(prec, scale)

        // List:
        | List innerType ->
            use innerHandle = innerType.CreateHandle()
            
            PolarsWrapper.NewListType innerHandle

        // Struct
        | Struct fields ->
            let names = fields |> List.map (fun f -> f.Name) |> List.toArray

            let typeHandles = fields |> List.map (fun f -> f.DataType.CreateHandle()) |> List.toArray
            
            try
                PolarsWrapper.NewStructType(names, typeHandles)
            finally
                for h in typeHandles do h.Dispose()

        | Array(innerType,width) ->
            use innerHandle = innerType.CreateHandle()
            PolarsWrapper.NewArrayType (innerHandle,width)

        | Unknown -> PolarsWrapper.NewPrimitiveType 0
        
/// <summary>
/// Represents the type of join operation to perform.
/// </summary>
type JoinType =
    | Inner
    | Left
    | Outer
    | Cross
    | Semi
    | Anti
    member internal this.ToNative() =
        match this with
        | Inner -> PlJoinType.Inner
        | Left -> PlJoinType.Left
        | Outer -> PlJoinType.Outer
        | Cross -> PlJoinType.Cross
        | Semi -> PlJoinType.Semi
        | Anti -> PlJoinType.Anti

/// <summary>
/// Specifies the aggregation function for pivot operations.
/// </summary>
type PivotAgg =
    | First | Sum | Min | Max | Mean | Median | Count | Last
    
    member internal this.ToNative() =
        match this with
        | First -> PlPivotAgg.First
        | Sum -> PlPivotAgg.Sum
        | Min -> PlPivotAgg.Min
        | Max -> PlPivotAgg.Max
        | Mean -> PlPivotAgg.Mean
        | Median -> PlPivotAgg.Median
        | Count -> PlPivotAgg.Count
        | Last -> PlPivotAgg.Last

/// <summary>
/// Specifies the type of concat operations.
/// </summary>
type ConcatType =
    | Vertical
    | Horizontal
    | Diagonal
    
    // 内部转换 helper
    member internal this.ToNative() =
        match this with
        | Vertical -> PlConcatType.Vertical
        | Horizontal -> PlConcatType.Horizontal
        | Diagonal -> PlConcatType.Diagonal

type Label =
    | Left 
    | Right 
    | DataPoint
    member internal this.ToNative() =
        match this with
        | Left -> PlLabel.Left
        | Right -> PlLabel.Right
        | DataPoint -> PlLabel.DataPoint 

type StartBy =
    | WindowBound
    | DataPoint
    | Monday
    | Tuesday
    | Wednesday
    | Thursday
    | Friday
    | Saturday
    | Sunday
    member internal this.ToNative() =
        match this with
        | WindowBound -> PlStartBy.WindowBound
        | DataPoint -> PlStartBy.DataPoint
        | Monday -> PlStartBy.Monday
        | Tuesday -> PlStartBy.Tuesday
        | Wednesday -> PlStartBy.Wednesday
        | Thursday -> PlStartBy.Thursday
        | Friday -> PlStartBy.Friday    
        | Saturday -> PlStartBy.Saturday
        | Sunday -> PlStartBy.Sunday

type ClosedWindow =
    | Left
    | Right
    | Both
    | NoWindow
    member internal this.ToNative() =
        match this with
        | Left -> PlClosedWindow.Left
        | Right -> PlClosedWindow.Right
        | Both -> PlClosedWindow.Both
        | NoWindow -> PlClosedWindow.None

type Roll =
    | Raise 
    | Forward 
    | Backward
    member internal this.ToNative() =
        match this with
        | Raise -> PlRoll.Raise
        | Forward -> PlRoll.Forward
        | Backward -> PlRoll.Backward

type QuantileMethod =
    | Nearest 
    | Higher 
    | Lower
    | Midpoint
    | Linear
    member internal this.ToNative() =
        match this with
        | Nearest -> PlQuantileMethod.Nearest
        | Higher -> PlQuantileMethod.Higher
        | Lower -> PlQuantileMethod.Lower
        | Midpoint -> PlQuantileMethod.Midpoint
        | Linear -> PlQuantileMethod.Linear

type RankMethod =
    | Average 
    | Min
    | Max
    | Dense
    | Ordinal
    | Random
    member internal this.ToNative() =
        match this with
        | Average -> PlRankMethod.Average
        | Min -> PlRankMethod.Min
        | Max -> PlRankMethod.Max
        | Dense -> PlRankMethod.Dense
        | Ordinal -> PlRankMethod.Ordinal
        | Random -> PlRankMethod.Random

type JoinValidation =
    | ManyToMany
    | ManyToOne
    | OneToMany
    | OneToOne
    member internal this.ToNative() =
        match this with
        | ManyToMany -> PlJoinValidation.ManyToMany
        | ManyToOne -> PlJoinValidation.ManyToOne
        | OneToMany -> PlJoinValidation.OneToMany
        | OneToOne -> PlJoinValidation.OneToOne

type JoinCoalesce =
    | JoinSpecific
    | CoalesceColumns
    | KeepColumns
    member internal this.ToNative() =
        match this with
        | JoinSpecific -> PlJoinCoalesce.JoinSpecific
        | CoalesceColumns -> PlJoinCoalesce.CoalesceColumns
        | KeepColumns -> PlJoinCoalesce.KeepColumns

type JoinMaintainOrder =
    | NotMaintainOrder
    | Left
    | Right
    | LeftRight
    | RightLeft
    member internal this.ToNative() =
        match this with
        | NotMaintainOrder -> PlJoinMaintainOrder.None
        | Left -> PlJoinMaintainOrder.Left
        | Right -> PlJoinMaintainOrder.Right
        | LeftRight -> PlJoinMaintainOrder.LeftRight
        | RightLeft -> PlJoinMaintainOrder.RightLeft

type AsofStrategy =
    | Backward
    | Forward
    | Nearest
    member internal this.ToNative() =
        match this with
        | Backward -> PlAsofStrategy.Backward
        | Forward -> PlAsofStrategy.Forward
        | Nearest -> PlAsofStrategy.Nearest

type ParallelStrategy =
    | Auto
    | Columns
    | RowGroups
    | NoParallel
    member internal this.ToNative() =
        match this with
        | Auto -> PlParallelStrategy.Auto
        | Columns -> PlParallelStrategy.Columns
        | RowGroups -> PlParallelStrategy.RowGroups
        | NoParallel -> PlParallelStrategy.None

File: Polars.CSharp/Enums.cs
#pragma warning disable CS1591
using CoreEnums = Polars.NET.Core;

namespace Polars.CSharp;
/// <summary>
/// Enums of JoinTypes
/// </summary>
public enum JoinType: byte
{
    Inner,Left, Outer,Cross,Semi,Anti,IEJoin
}
/// <summary>
/// Specifies the aggregation function for pivot operations.
/// </summary>
public enum PivotAgg : byte
{
    First,Sum,Min,Max, Mean,Median,Count,Len,Last
}
/// <summary>
/// TimeUnit Enums
/// </summary>
public enum TimeUnit
{
    Nanoseconds = 0,
    Microseconds = 1,
    Milliseconds = 2,
    Second = 3,
    Minute = 4,
    Hour = 5,
    Day = 6,
    Month = 7,
    Year = 8
}
/// <summary>
/// Concat Type Enum
/// </summary>
public enum ConcatType
{
    Vertical,Horizontal,Diagonal
}

/// <summary>
/// Enum of DataTypeKind
/// </summary>
public enum DataTypeKind
{
    Boolean = 1 ,
    Int8 =2,
    Int16=3,
    Int32=4,
    Int64=5,
    UInt8=6,
    UInt16=7,
    UInt32=8,UInt64=9,
    Float32=10,Float64=11,
    String=12,Date=13,Datetime=14,Time=15,Duration=16,
    Binary=17,
    Null=18,
    Struct=19,List=20,Categorical=21,
    Decimal=22,
    Array=23,
    Int128 = 24,
    UInt128=25,
    Unknown = 0,
    SameAsInput=0
}

/// <summary>
/// Defines which boundary of the window to use for the label.
/// </summary>
public enum Label
{
    Left,
    Right,
    DataPoint
}

/// <summary>
/// Defines the strategy for determining the start of the window.
/// </summary>
public enum StartBy
{
    WindowBound,
    DataPoint,
    Monday,
    Tuesday,
    Wednesday,
    Thursday,
    Friday,
    Saturday,
    Sunday
}

/// <summary>
/// Defines which window boundaries are closed (inclusive).
/// </summary>
public enum ClosedWindow
{
    Left,
    Right,
    Both,
    None
}

/// <summary>
/// Strategy to handle dates that land on non-business days (weekends or holidays).
/// </summary>
public enum Roll
{
    /// <summary>
    /// Raise an error if the result is not a business day.
    /// </summary>
    Raise,
    
    /// <summary>
    /// Roll forward to the next business day.
    /// </summary>
    Forward,
    
    /// <summary>
    /// Roll backward to the previous business day.
    /// </summary>
    Backward
}

public enum QuantileMethod : byte
{
    Nearest = 0,
    Higher = 1,
    Lower = 2,
    Midpoint = 3,
    Linear = 4 // Default
}

public enum RankMethod : byte
{
    /// <summary>
    ///  The average of the ranks that would have been assigned to all the tied values is assigned to each value.
    /// </summary>
    Average = 0, // Default
    /// <summary>
    /// The minimum of the ranks that would have been assigned to all the tied values is assigned to each value. (This is also referred to as “competition” ranking.)
    /// </summary>
    Min = 1,
    /// <summary>
    /// The maximum of the ranks that would have been assigned to all the tied values is assigned to each value.
    /// </summary>
    Max = 2,
    /// <summary>
    /// Like ‘min’, but the rank of the next highest element is assigned the rank immediately after those assigned to the tied elements.
    /// </summary>
    Dense = 3,
    /// <summary>
    /// All values are given a distinct rank, corresponding to the order that the values occur in the Series.
    /// </summary>
    Ordinal = 4,
    /// <summary>
    /// Like ‘ordinal’, but the rank for ties is not dependent on the order that the values occur in the Series.
    /// </summary>
    Random = 5
}

public enum RollingRankMethod : byte
{
    /// <summary>
    ///  The average of the ranks that would have been assigned to all the tied values is assigned to each value.
    /// </summary>
    Average = 0, // Default
    /// <summary>
    /// The minimum of the ranks that would have been assigned to all the tied values is assigned to each value. (This is also referred to as “competition” ranking.)
    /// </summary>
    Min = 1,
    /// <summary>
    /// The maximum of the ranks that would have been assigned to all the tied values is assigned to each value.
    /// </summary>
    Max = 2,
    /// <summary>
    /// Like ‘min’, but the rank of the next highest element is assigned the rank immediately after those assigned to the tied elements.
    /// </summary>
    Dense = 3,
    /// <summary>
    /// All values are given a distinct rank, corresponding to the order that the values occur in the Series.
    /// </summary>
    Random = 4
}

public enum UniqueKeepStrategy : byte
{
    First = 0,
    Last = 1,
    Any = 2,
    None = 3
}

public enum JoinValidation: byte
{
    // Default
    ManyToMany = 0,
    ManyToOne = 1,
    OneToMany = 2,
    OneToOne = 3
}

public enum JoinCoalesce: byte
{
    // Default
    JoinSpecific = 0,
    CoalesceColumns = 1,
    KeepColumns = 2,
}

public enum JoinMaintainOrder: byte
{
    // Default
    None = 0,
    Left = 1,
    Right =2,
    LeftRight =3,
    RightLeft =4
}

public enum AsofStrategy: byte
{
    Backward = 0,
    Forward = 1,
    Nearest = 2
}

public enum ParallelStrategy: byte
{
    Auto = 0,
    Columns = 1,
    RowGroups = 2,
    None =3
}

public enum CsvEncoding: byte
{
    UTF8 = 0,
    LossyUTF8 = 1,
}

public enum JsonFormat: byte
{
    Json = 0,
    JsonLines = 1,
}

public enum IpcCompression: byte
{
    None = 0,
    LZ4 = 1,
    ZSTD = 2
}

/// <summary>
/// Controls the file synchronization behavior when closing the file.
/// </summary>
public enum SyncOnClose : byte
{
    /// <summary>
    /// Don't call sync on close. (Default, fastest)
    /// </summary>
    None = 0,
    
    /// <summary>
    /// Sync only the file contents.
    /// </summary>
    Data = 1,
    
    /// <summary>
    /// Sync the file contents and the metadata. (Slowest, safest)
    /// </summary>
    All = 2
}

public enum ParquetCompression : byte
{
    None = 0,
    Snappy = 1,
    Gzip = 2,
    Brotli = 3,
    Zstd = 4,
    Lz4Raw = 5
}

public enum QuoteStyle : byte
{
    /// <summary>
    /// Quote fields only when necessary (e.g. containing delimiter/quote). Default.
    /// </summary>
    Necessary = 0,
    
    /// <summary>
    /// Quote every field.
    /// </summary>
    Always = 1,
    
    /// <summary>
    /// Quote non-numeric fields.
    /// </summary>
    NonNumeric = 2,
    
    /// <summary>
    /// Never quote fields.
    /// </summary>
    Never = 3
}

public enum InterpolationMethod
{
    Linear = 0,
    Nearest = 1
}

/// <summary>
/// Supported cloud providers for IO operations.
/// </summary>
public enum CloudProvider : byte
{
    /// <summary>
    /// No cloud provider / Local file.
    /// </summary>
    None = 0,

    /// <summary>
    /// Amazon S3 or compatible services (MinIO, etc).
    /// </summary>
    Aws = 1,

    /// <summary>
    /// Azure Blob Storage / Data Lake Gen2.
    /// </summary>
    Azure = 2,

    /// <summary>
    /// Google Cloud Storage.
    /// </summary>
    Gcp = 3,

    /// <summary>
    /// Generic HTTP/HTTPS.
    /// </summary>
    Http = 4,

    /// <summary>
    /// Hugging Face datasets.
    /// </summary>
    HuggingFace = 5
}

internal static class EnumExtensions
{
    public static CoreEnums.PlDataType ToNative(this DataTypeKind kind) => kind switch
    {
        DataTypeKind.SameAsInput => CoreEnums.PlDataType.SameAsInput,
        DataTypeKind.Int8 => CoreEnums.PlDataType.Int8,
        DataTypeKind.UInt8 => CoreEnums.PlDataType.UInt8,
        DataTypeKind.Int16 => CoreEnums.PlDataType.Int16,
        DataTypeKind.UInt16 => CoreEnums.PlDataType.UInt16,
        DataTypeKind.Int32=> CoreEnums.PlDataType.Int32,
        DataTypeKind.UInt64 => CoreEnums.PlDataType.UInt64,
        DataTypeKind.Int128 => CoreEnums.PlDataType.Int128,
        DataTypeKind.UInt128 => CoreEnums.PlDataType.UInt128,
        DataTypeKind.Float32 => CoreEnums.PlDataType.Float32,
        DataTypeKind.Float64 => CoreEnums.PlDataType.Float64,
        DataTypeKind.Datetime => CoreEnums.PlDataType.Datetime,
        DataTypeKind.Date => CoreEnums.PlDataType.Date,
        DataTypeKind.Time => CoreEnums.PlDataType.Time,
        DataTypeKind.Duration => CoreEnums.PlDataType.Duration,
        DataTypeKind.List=> CoreEnums.PlDataType.List,
        DataTypeKind.Array => CoreEnums.PlDataType.Array,
        DataTypeKind.Struct => CoreEnums.PlDataType.Struct,
        DataTypeKind.Binary => CoreEnums.PlDataType.Binary,
        DataTypeKind.Decimal => CoreEnums.PlDataType.Decimal,
        DataTypeKind.Boolean => CoreEnums.PlDataType.Boolean,
        DataTypeKind.Categorical => CoreEnums.PlDataType.Categorical,
        DataTypeKind.Null => CoreEnums.PlDataType.Null,
        DataTypeKind.String => CoreEnums.PlDataType.String, 
        _ => throw new ArgumentOutOfRangeException(nameof(kind), kind, null)
    };
    public static CoreEnums.PlTimeUnit ToNative(this TimeUnit unit) => unit switch
    {
        TimeUnit.Nanoseconds => CoreEnums.PlTimeUnit.Nanoseconds,
        TimeUnit.Microseconds => CoreEnums.PlTimeUnit.Microseconds,
        TimeUnit.Milliseconds => CoreEnums.PlTimeUnit.Milliseconds,
        TimeUnit.Second => CoreEnums.PlTimeUnit.Second,
        TimeUnit.Minute => CoreEnums.PlTimeUnit.Minute,
        TimeUnit.Hour => CoreEnums.PlTimeUnit.Hour,
        TimeUnit.Day => CoreEnums.PlTimeUnit.Day,
        TimeUnit.Month => CoreEnums.PlTimeUnit.Month,
        TimeUnit.Year => CoreEnums.PlTimeUnit.Year,
        _ => CoreEnums.PlTimeUnit.Nanoseconds
    };
    public static CoreEnums.PlJoinType ToNative(this JoinType type) => type switch
    {
        JoinType.Inner => CoreEnums.PlJoinType.Inner,
        JoinType.Left => CoreEnums.PlJoinType.Left,
        JoinType.Outer => CoreEnums.PlJoinType.Outer,
        JoinType.Cross => CoreEnums.PlJoinType.Cross,
        JoinType.Semi => CoreEnums.PlJoinType.Semi,
        JoinType.Anti => CoreEnums.PlJoinType.Anti,
        JoinType.IEJoin => CoreEnums.PlJoinType.IEJoin,
        _ => CoreEnums.PlJoinType.Inner
    };

    public static CoreEnums.PlPivotAgg ToNative(this PivotAgg agg) => agg switch
    {
        PivotAgg.First => CoreEnums.PlPivotAgg.First,
        PivotAgg.Sum => CoreEnums.PlPivotAgg.Sum,
        PivotAgg.Min => CoreEnums.PlPivotAgg.Min,
        PivotAgg.Max => CoreEnums.PlPivotAgg.Max,
        PivotAgg.Mean => CoreEnums.PlPivotAgg.Mean,
        PivotAgg.Median => CoreEnums.PlPivotAgg.Median,
        PivotAgg.Count => CoreEnums.PlPivotAgg.Count,
        PivotAgg.Len => CoreEnums.PlPivotAgg.Len,
        PivotAgg.Last => CoreEnums.PlPivotAgg.Last,
        _ => CoreEnums.PlPivotAgg.First
    };
    
    public static CoreEnums.PlConcatType ToNative(this ConcatType type) => type switch
    {
        ConcatType.Vertical => CoreEnums.PlConcatType.Vertical,
        ConcatType.Horizontal => CoreEnums.PlConcatType.Horizontal,
        ConcatType.Diagonal => CoreEnums.PlConcatType.Diagonal,
        _ => CoreEnums.PlConcatType.Vertical
    };
    internal static CoreEnums.PlLabel ToNative(this Label label) => label switch
    {
        Label.Left => CoreEnums.PlLabel.Left,
        Label.Right => CoreEnums.PlLabel.Right,
        Label.DataPoint => CoreEnums.PlLabel.DataPoint,
        _ => throw new ArgumentOutOfRangeException(nameof(label), label, null)
    };

    internal static CoreEnums.PlStartBy ToNative(this StartBy startBy) => startBy switch
    {
        StartBy.WindowBound => CoreEnums.PlStartBy.WindowBound,
        StartBy.DataPoint => CoreEnums.PlStartBy.DataPoint,
        StartBy.Monday => CoreEnums.PlStartBy.Monday,
        StartBy.Tuesday => CoreEnums.PlStartBy.Tuesday,
        StartBy.Wednesday => CoreEnums.PlStartBy.Wednesday,
        StartBy.Thursday => CoreEnums.PlStartBy.Thursday,
        StartBy.Friday => CoreEnums.PlStartBy.Friday,
        StartBy.Saturday => CoreEnums.PlStartBy.Saturday,
        StartBy.Sunday => CoreEnums.PlStartBy.Sunday,
        _ => throw new ArgumentOutOfRangeException(nameof(startBy), startBy, null)
    };

    internal static CoreEnums.PlClosedWindow ToNative(this ClosedWindow closed) => closed switch
    {
        ClosedWindow.Left => CoreEnums.PlClosedWindow.Left,
        ClosedWindow.Right => CoreEnums.PlClosedWindow.Right,
        ClosedWindow.Both => CoreEnums.PlClosedWindow.Both,
        ClosedWindow.None => CoreEnums.PlClosedWindow.None,
        _ => throw new ArgumentOutOfRangeException(nameof(closed), closed, null)
    };

    internal static CoreEnums.PlRoll ToNative(this Roll roll) => roll switch
    {
        Roll.Backward => CoreEnums.PlRoll.Backward,
        Roll.Forward => CoreEnums.PlRoll.Forward,
        Roll.Raise => CoreEnums.PlRoll.Raise,
        _ => throw new ArgumentOutOfRangeException(nameof(roll), roll, null)
    };
    internal static CoreEnums.PlQuantileMethod ToNative(this QuantileMethod interpol) => interpol switch
    {
        QuantileMethod.Nearest => CoreEnums.PlQuantileMethod.Nearest,
        QuantileMethod.Higher => CoreEnums.PlQuantileMethod.Higher,
        QuantileMethod.Lower => CoreEnums.PlQuantileMethod.Lower,
        QuantileMethod.Midpoint => CoreEnums.PlQuantileMethod.Midpoint,
        QuantileMethod.Linear => CoreEnums.PlQuantileMethod.Linear,  
        _ => throw new ArgumentOutOfRangeException(nameof(interpol), interpol, null)
    };
    internal static CoreEnums.PlRankMethod ToNative(this RankMethod method) => method switch
    {
        RankMethod.Average => CoreEnums.PlRankMethod.Average,
        RankMethod.Min => CoreEnums.PlRankMethod.Min,
        RankMethod.Max => CoreEnums.PlRankMethod.Max,
        RankMethod.Dense => CoreEnums.PlRankMethod.Dense,
        RankMethod.Ordinal => CoreEnums.PlRankMethod.Ordinal,
        RankMethod.Random => CoreEnums.PlRankMethod.Random,
        _ => throw new ArgumentOutOfRangeException(nameof(method), method, null)
    };
    internal static CoreEnums.PlRollingRankMethod ToNative(this RollingRankMethod method) => method switch
    {
        RollingRankMethod.Average => CoreEnums.PlRollingRankMethod.Average,
        RollingRankMethod.Min => CoreEnums.PlRollingRankMethod.Min,
        RollingRankMethod.Max => CoreEnums.PlRollingRankMethod.Max,
        RollingRankMethod.Dense => CoreEnums.PlRollingRankMethod.Dense,
        RollingRankMethod.Random => CoreEnums.PlRollingRankMethod.Random,
        _ => throw new ArgumentOutOfRangeException(nameof(method), method, null)
    };
    internal static CoreEnums.PlUniqueKeepStrategy ToNative(this UniqueKeepStrategy strategy) => strategy switch
    {
        UniqueKeepStrategy.First => CoreEnums.PlUniqueKeepStrategy.First,
        UniqueKeepStrategy.Last => CoreEnums.PlUniqueKeepStrategy.Last,
        UniqueKeepStrategy.Any => CoreEnums.PlUniqueKeepStrategy.Any,
        UniqueKeepStrategy.None => CoreEnums.PlUniqueKeepStrategy.None,
        _ => throw new ArgumentOutOfRangeException(nameof(strategy), strategy, null)
    };
    internal static CoreEnums.PlJoinValidation ToNative(this JoinValidation validation) => validation switch
    {
        JoinValidation.ManyToMany => CoreEnums.PlJoinValidation.ManyToMany,
        JoinValidation.ManyToOne => CoreEnums.PlJoinValidation.ManyToOne,
        JoinValidation.OneToMany => CoreEnums.PlJoinValidation.OneToMany,
        JoinValidation.OneToOne => CoreEnums.PlJoinValidation.OneToOne,
        _ => throw new ArgumentOutOfRangeException(nameof(validation), validation, null)
    };
    internal static CoreEnums.PlJoinCoalesce ToNative(this JoinCoalesce coalesce) => coalesce switch
    {
        JoinCoalesce.JoinSpecific => CoreEnums.PlJoinCoalesce.JoinSpecific,
        JoinCoalesce.CoalesceColumns => CoreEnums.PlJoinCoalesce.CoalesceColumns,
        JoinCoalesce.KeepColumns => CoreEnums.PlJoinCoalesce.KeepColumns,
        _ => throw new ArgumentOutOfRangeException(nameof(coalesce), coalesce, null)
    };
    internal static CoreEnums.PlJoinMaintainOrder ToNative(this JoinMaintainOrder maintainOrder) => maintainOrder switch
    {
        JoinMaintainOrder.None => CoreEnums.PlJoinMaintainOrder.None,
        JoinMaintainOrder.Left => CoreEnums.PlJoinMaintainOrder.Left,
        JoinMaintainOrder.Right => CoreEnums.PlJoinMaintainOrder.Right,
        JoinMaintainOrder.LeftRight => CoreEnums.PlJoinMaintainOrder.LeftRight,
        JoinMaintainOrder.RightLeft => CoreEnums.PlJoinMaintainOrder.RightLeft,
        _ => throw new ArgumentOutOfRangeException(nameof(maintainOrder), maintainOrder, null)
    };
    internal static CoreEnums.PlAsofStrategy ToNative(this AsofStrategy strategy) => strategy switch
    {
        AsofStrategy.Backward => CoreEnums.PlAsofStrategy.Backward,
        AsofStrategy.Forward => CoreEnums.PlAsofStrategy.Forward,
        AsofStrategy.Nearest => CoreEnums.PlAsofStrategy.Nearest,
        _ => throw new ArgumentOutOfRangeException(nameof(strategy), strategy, null)
    };
    internal static CoreEnums.PlParallelStrategy ToNative(this ParallelStrategy strategy) => strategy switch
    {
        ParallelStrategy.Auto => CoreEnums.PlParallelStrategy.Auto,
        ParallelStrategy.Columns => CoreEnums.PlParallelStrategy.Columns,
        ParallelStrategy.RowGroups => CoreEnums.PlParallelStrategy.RowGroups,
        ParallelStrategy.None => CoreEnums.PlParallelStrategy.None,
        _ => throw new ArgumentOutOfRangeException(nameof(strategy), strategy, null)
    };
    internal static CoreEnums.PlCsvEncoding ToNative(this CsvEncoding encoding) => encoding switch
    {
        CsvEncoding.UTF8 => CoreEnums.PlCsvEncoding.UTF8,
        CsvEncoding.LossyUTF8 => CoreEnums.PlCsvEncoding.LossyUTF8,
        _ => throw new ArgumentOutOfRangeException(nameof(encoding), encoding, null)
    };
    internal static CoreEnums.PlJsonFormat ToNative(this JsonFormat jsonFormat) => jsonFormat switch
    {
        JsonFormat.Json => CoreEnums.PlJsonFormat.Json,
        JsonFormat.JsonLines => CoreEnums.PlJsonFormat.JsonLines,
        _ => throw new ArgumentOutOfRangeException(nameof(jsonFormat), jsonFormat, null)
    };
    internal static CoreEnums.PlIpcCompression ToNative(this IpcCompression compression) => compression switch
    {
        IpcCompression.None => CoreEnums.PlIpcCompression.None,
        IpcCompression.LZ4 => CoreEnums.PlIpcCompression.LZ4,
        IpcCompression.ZSTD => CoreEnums.PlIpcCompression.ZSTD,
        _ => throw new ArgumentOutOfRangeException(nameof(compression), compression, null)
    };
    internal static CoreEnums.PlSyncOnClose ToNative(this SyncOnClose syncOnClose) => syncOnClose switch
    {
        SyncOnClose.None => CoreEnums.PlSyncOnClose.None,
        SyncOnClose.Data => CoreEnums.PlSyncOnClose.Data,
        SyncOnClose.All => CoreEnums.PlSyncOnClose.All,
        _ => throw new ArgumentOutOfRangeException(nameof(syncOnClose), syncOnClose, null)
    };
    internal static CoreEnums.PlParquetCompression ToNative(this ParquetCompression compression) => compression switch
    {
        ParquetCompression.None => CoreEnums.PlParquetCompression.Uncompressed,
        ParquetCompression.Snappy => CoreEnums.PlParquetCompression.Snappy,
        ParquetCompression.Gzip => CoreEnums.PlParquetCompression.Gzip,
        ParquetCompression.Brotli => CoreEnums.PlParquetCompression.Brotli,
        ParquetCompression.Zstd => CoreEnums.PlParquetCompression.Zstd,
        ParquetCompression.Lz4Raw => CoreEnums.PlParquetCompression.Lz4Raw,
        _ => throw new ArgumentOutOfRangeException(nameof(compression), compression, null)
    };
    internal static CoreEnums.PlQuoteStyle ToNative(this QuoteStyle style) => style switch
    {
        QuoteStyle.Necessary => CoreEnums.PlQuoteStyle.Necessary,
        QuoteStyle.Always => CoreEnums.PlQuoteStyle.Always,
        QuoteStyle.NonNumeric => CoreEnums.PlQuoteStyle.NonNumeric,
        QuoteStyle.Never => CoreEnums.PlQuoteStyle.Never,
        _ => throw new ArgumentOutOfRangeException(nameof(style), style, null)
    };
    internal static CoreEnums.PlInterpolationMethod ToNative(this InterpolationMethod method) => method switch
    {
        InterpolationMethod.Nearest => CoreEnums.PlInterpolationMethod.Nearest,
        InterpolationMethod.Linear => CoreEnums.PlInterpolationMethod.Linear,
        _ => throw new ArgumentOutOfRangeException(nameof(method), method, null)
    };
    internal static CoreEnums.PlCloudProvider ToNative(this CloudProvider cloud) => cloud switch
    {
        CloudProvider.None => CoreEnums.PlCloudProvider.None,
        CloudProvider.Aws => CoreEnums.PlCloudProvider.Aws,
        CloudProvider.Azure => CoreEnums.PlCloudProvider.Azure,
        CloudProvider.Gcp => CoreEnums.PlCloudProvider.Gcp,
        CloudProvider.Http => CoreEnums.PlCloudProvider.Http,
        CloudProvider.HuggingFace => CoreEnums.PlCloudProvider.HuggingFace,
        _ => throw new ArgumentOutOfRangeException(nameof(cloud), cloud, null)
    };
}

